feat(iota-types, iota-execution): Introduction of new basic types (ChangeEpochTransactionV4)

oliviasaa opened the pull request:
```mermaid
---
config:
  gitGraph:
    {parallelCommits: true,
    mainBranchName: 'Feature',
    rotateCommitLabel: true}
---
gitGraph
   commit id: "Initial commit"
   commit id: "Old Scorer" tag: "PR8530"
   checkout Feature
      branch new-types
   commit id: "add ChangeEpochV4"
   checkout Feature
    branch misbehavior-reports
   commit id: "add MisbehaviorReports"
    checkout misbehavior-reports
    branch scoring-metrics
   commit id: "add ScoringMetrics"
    checkout scoring-metrics
   branch scorer
   commit id: "add Scorer"
    checkout scorer
   branch scoring-metrics-store
   commit id: "add ScoringMetricsStore"
branch report-validation
commit id: "add report validation" 
branch scoring-function
commit id: "add scoring function" 
branch report-creation
commit id: "add report creation" 
branch advance-epoch
commit id: "add advance epoch v4" 
 checkout Feature
    merge new-types tag: "THIS PR"
```
# Description of change

This PR introduces: 
- a new variant of `EndOfEpochTransactionKind` which includes the `scores` field. This transaction is used when score based rewards are enabled (i.e., whenever the newly added feature flag `score_based_rewards` is set to true). 
- a placeholder for the new `advance_epoch_v4` function, currently with the same logic as `advance_epoch_v3`
- related changes in `iota-graphql-rpc` crate, as the `ChangeEpochTransactionV4` type, and a new variant of `IotaEndOfEpochTransactionKind`.

Note: the changes in the PR are built on top of commit [c7eeee2](https://github.com/iotaledger/iota-rust-sdk/tree/feat/add-scores) of `iota-rust-sdk`, which is currently not even merged to `develop`. Thus, any changes present in this PR should be merged to `iota`'s `develop` branch together with the changes in `iota-rust-sdk` repo.

## How the change has been tested

- [x] Basic tests (linting, compilation, formatting, unit/integration tests)
- [ ] Patch-specific tests (correctness, functionality coverage)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that new and existing unit tests pass locally with my changes

### Release Notes

- [ ] Protocol:
- [x] Nodes (Validators and Full nodes): Introduce type ChangeEpochTransactionV4
- [ ] Indexer:
- [ ] JSON-RPC:
- [ ] GraphQL:
- [ ] CLI:
- [ ] Rust SDK:
- [ ] REST API:

oliviasaa committed:
feat(iota-graphql-rpc, iota-json-rpc-types, iota-types, iota-execution): Introduction of ChangeEpochV4 transaction (#8522)

Introduces the ChangeEpochV4 type, which includes a vector of scores
relative to each authority behavior in the ending epoch.

- [x] Basic tests (linting, compilation, formatting, unit/integration
tests)
- [ ] Patch-specific tests (correctness, functionality coverage)
- [ ] I have added tests that prove my fix is effective or that my
feature works
- [x] I have checked that new and existing unit tests pass locally with
my changes

- [ ] Protocol:
- [x] Nodes (Validators and Full nodes): Introduces the ChangeEpochV4
type
- [ ] Indexer:
- [ ] JSON-RPC:
- [ ] GraphQL:
- [ ] CLI:
- [ ] Rust SDK:
- [ ] REST API:

oliviasaa committed:
apply upstream changes

oliviasaa committed:
Initial commit

oliviasaa committed:
add feature flag and advance_epoch_v4 placeholder

Thoralf-M commented on crates/iota-graphql-rpc/src/types/transaction_block_kind/end_of_epoch.rs:
```diff
@@ -47,6 +48,7 @@ pub(crate) struct EndOfEpochTransaction {
 pub(crate) enum EndOfEpochTransactionKind {
     ChangeEpoch(ChangeEpochTransaction),
     ChangeEpochV2(ChangeEpochTransactionV2),
+    ChangeEpochV4(ChangeEpochTransactionV4),
```
Why is there no V3?

Thoralf-M replied:
Because V2 has the fields for V3 optional added https://github.com/iotaledger/iota/pull/9232/files#r2517585873

oliviasaa committed:
generate format

cyberphysic4l commented on crates/iota-protocol-config/src/lib.rs:
```diff
@@ -2294,6 +2303,10 @@ impl ProtocolConfig {
                 }
+                16 => {
+                    // Enables score based rewards.
+                    cfg.feature_flags.score_based_rewards = true;
```
We should presumably enable this on devnet to start with

cyberphysic4l commented on crates/iota-types/src/transaction.rs:
```diff
@@ -269,6 +269,41 @@ pub struct ChangeEpochV3 {
+    /// Vector of active validator indices eligible to take part in committee
+    /// selection because they support the new, target protocol version.
+    pub eligible_active_validators: Vec<u64>,
+    /// Scores relative to the epoch being finalized. Each value corresponds to
```
```suggestion
    /// Scores for the epoch being finalized. Each value corresponds to
```

cyberphysic4l commented on crates/iota-types/src/transaction.rs:
```diff
@@ -567,6 +641,23 @@ impl EndOfEpochTransactionKind {
+                        "selecting committee only among validators supporting the protocol version required".to_string(),
+                    ));
+                }
+                if !config.score_based_rewards() {
+                    return Err(UserInputError::Unsupported(
+                        "score based rewards not enabled".to_string(),
+                    ));
+                }
```
Need to check this is not enable for all previous version above as well.

kodemartin reviewed (commented):
Prepared a patch with #9232 to extend existing types in the web APIs (graphql, json-rpc) as opposed to introducing new types. 

This follows the pattern we used for `ChangeEpochV3` and ensures that the Rust API remains minimal, and that the schema remains backward compatible.

cyberphysic4l reviewed (commented):
I am happy with @kodemartin 's patch. 
Also, the parts that are done seem fine, but there is a good bit left to do. You have listed most remaining stuff in the PR description, but this must include the changes in `iota-framework` for the `advance_epoch` function there too and all the required changes with new snapshots etc.

kodemartin committed:
refactor(graphql-rpc,json-rpc): extend existing types to accommodate protocol ChangeEpochV4 (#9232)

# Description of change

This follows #9159 and refactors `graphql-` and `json-rpc` types, by
adding the new `scores` field as an optional field to the API types.

## How the change has been tested

Describe the tests that you ran to verify your changes.

Make sure to provide instructions for the maintainer as well as any
relevant configurations.

- [x] Basic tests (linting, compilation, formatting, unit/integration
tests)

kodemartin reviewed (approved):
lgtm for the infra part

oliviasaa committed:
enables score based rewards in Devnet only

oliviasaa commented:
Obs: part of the [validity check](https://github.com/iotaledger/iota/blob/dd187d17bea5eba1701964917b16137661b14130/crates/iota-types/src/transaction.rs#L606) was commented out because those messages are still not being sent in this current PR, which makes the tests break. Adding those checks was already added as a todo to the description of the feature branch

oliviasaa committed:
Generate schemas and snapshot files

piotrm50 commented on crates/iota-types/src/transaction.rs:
```diff
@@ -381,6 +416,7 @@ pub enum EndOfEpochTransactionKind {
     ChangeEpoch(ChangeEpoch),
     ChangeEpochV2(ChangeEpochV2),
     ChangeEpochV3(ChangeEpochV3),
+    ChangeEpochV4(ChangeEpochV4),
```
Add the enum value at the end of the list to avoid potential serialization/deserialization problems.

oliviasaa replied:
Done

oliviasaa committed:
[skip ci] fix ordering of values in enum EndOfEpochTransactionKind
